stdlib/runtime/runtimeDebug.gr

```diff
@@ -0,0 +1,712 @@
+import NumberUtils from "runtime/numberUtils"
+import Map from "map"
+import Set from "set"
+import Option from "option"
+import Tags from "runtime/unsafe/tags"
+
+@disableGC
+let rec toHex = (n: Number) => {
+  let x = coerceNumberToWasmI32(n)
+  let result = NumberUtils.itoa32(x, 16n)
+
+  Memory.decRef(WasmI32.fromGrain(n))
+  Memory.decRef(WasmI32.fromGrain(toHex))
+
+  result
+}
+
+export enum LeakDebugSummaryMode {
+  AlwaysPrintSummary,
+  PrintSummaryIfLeaksAreFound,
+}
+
+export record LeakDebugBlockOptions {
+  printIncrefsAndDecRefs: Bool,
+  printMallocAndFree: Bool,
+  printObjectContents: Bool,
+  summary: LeakDebugSummaryMode,
+}
+
+let incrementCountInMap = (key: k, map: Map.Map<k, Number>) => {
+  match (Map.get(key, map)) {
+    Some(cnt) => Map.set(key, cnt + 1, map),
+    None => Map.set(key, 1, map),
+  }
+}
+
+let keyUnionOfTwoMaps = (map1: Map.Map<k, v1>, map2: Map.Map<k, v2>) => {
+  let keyUnion = Set.make(): (Set.Set<k>)
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map1,
+  )
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map2,
+  )
+
+  keyUnion
+}
+
+let leakDebugBlockImpl = (options: LeakDebugBlockOptions, f: () -> Void) => {
+  // The main objective is to find and print objects allocated
+  // in the function, but not deallocated at the end.
+
+  let allocations = Set.make(): (Set.Set<Number>)
+
+  let tagToTypeName = (tag: Number) => {
+    match (tag) {
+      1 => "string",
+      2 => "char",
+      3 => "enum",
+      4 => "record",
+      5 => "array",
+      6 => "number",
+      7 => "lambda",
+      8 => "tuple",
+      9 => "bytes",
+      _ => "unknown",
+    }
+  }
+
+  let mut totalMallocs = 0
+  let mut totalFrees = 0
+  let mut totalIncRefs = 0
+  let mut totalDecRefs = 0
+
+  let mut totalPreexistingObjectsDeallocated = 0
+
+  let incRefsByTag = Map.make(): (Map.Map<Number, Number>)
+  let decRefsByTag = Map.make(): (Map.Map<Number, Number>)
+
+  let inintialRefCountsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+  let incRefsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+  let decRefsOnPreExistingObjects = Map.make(): (Map.Map<Number, Number>)
+
+  // Since objects created by GC.malloc have a starting refcount of 1,
+  // the total incRefsByTag would not match decRefsByTag. I'm compensating it
+  // by keeping track of number of objects that have gone malloc+free cycle.
+  // This isn't perfect because in case of leaks free isn't called and the
+  // totals can create confusion. The simplest solution would be to get the tag
+  // of mallocated objects, but they don't yet have a tag associated. So I
+  // have to compensate again by adding up the objects that have been
+  // mallocated but never freed. Which simply means the ones in "allocations"
+  // Map.
+  let mallocPlusFreeByTag = Map.make(): (Map.Map<Number, Number>)
+
+  @disableGC
+  let rec readTag = (userPtrAsNumber: Number) => {
```

Comment:
I can imagine why this is the case, but it may be worth a comment explaining why these functions use `Number` instead of `WasmI32`.

Reply:
I've added a comment explaining the choice for Number.